keps/sig-api-machinery/2839-in-use-protection/README.md

```diff
@@ -0,0 +1,761 @@
+
+#### Race of removing/adding liens
+
+A cluster-admin would be racing to update the object to remove the lien and delete the object before a namespace editor is able to place the lien back. As a result, a namespace admin or editor can create an object that a cluster-admin or namespace lifecycle controller cannot delete, which shouldn't be allowed.
+
+To mitigate the risk, a new `IgnoreLiens` API option in `DeleteOptions` to force delete a resource with liens will be added.
+
+Specifying the option should only be allowed to a limited set of users and groups, therefore deletion request will be blocked in api-server, if invalid users request deletion with the option.
+Restricting the `IgnoreLiens` API option to specific users is beta blocker and won't be included in the initial implementation.
+Therefore, in the initial implementation, any users with delete permission will be able to delete resources with liens by specifying the `IgnoreLiens` API option.
+
+#### Subresource
+
+There is a use case to disallow a controller to write to its CRD except for the status subresource, but allow it to update finalizers/liens.
+To allow this use case, it is required to provide a feature to set a separate RBAC to update finalizers/liens.
+On the other hand, finalizers field is currently not subresourced.
+Therefore, it needs to provide this feature in a compatible way for finalizers, and it would be better to implement liens field in the same way to finalizers.
+
+To solve this issue, there is [another KEP](https://github.com/kubernetes/enhancements/issues/3285) to add a new knob for just patching only a specific field via an endpoint of a subresource, without modifying the field itself.
+In this way, a feature to set a separate RBAC can be implemented later without making liens field itself a subresource.
+
+Subresourcing `Liens` is a beta blocker and won't be included in the initial implementation.
+
+#### Unresolved issues
+
+- Decision: When you delete a namespace and an object in that namespace has a lien:
+  - Does the namespace and object get deleted anyway?
+    - [x] Yes.
+  - Does the namespace deletion get blocked (i.e., the delete request fails) if any object in the namespace has a lien?
+    - [x] No.
+  - Does the other objects in the namespace get deleted, but that not object, preventing the namespace deletion from completing?
+    - [x] No. If you delete the namespace, the content will be removed.
+  - Does the namespace deletion not proceed until all objects in the namespace are free of liens?
+    - [ ] No. If you delete the namespace, the content will be removed. Namespace lifecycle cleanup will not honor liens.
+    - [x] No. If you delete the namespace, the content will be removed. Namespace lifecycle cleanup will honor liens only to guarantee the deletion order inside a namespace (guarantee of the deletion order is a beta blocker).
```

Comment:
I'm not sure deletion order is important UNLESS we can ultimately leave the deletion pending

Reply:
I also prefer to make deletion of resources with lien pending as an initial implementation (in this case, how cluster admin force the deletion of namespace will be a beta blocker), but either way is fine to me as long as the final goal is the same.

Reply:
> I'm not sure deletion order is important UNLESS we can ultimately leave the deletion pending

I think eventually deleting is important, but I can also imagine putting a timeframe on it (an hour?).  the namespace lifecycle controller could try to delete everything every few minutes and eventually, by blind luck, it will eventually make enough progress if an ordering exists.  After the timeout, the namespacelifecycle controller can override what remains.

The code that relies on finalizers today still needs to keep those finalizers in order to guarantee the cleanup.  Liens just make it less likely to have deletion in a "bad" order.

Reply:
>I think that we still have choices:
>to make an hour time out happen unconditionally

unconditionally is better.